File: solutions/haskell/zebra-puzzle/1/src/ZebraPuzzle.hs
module ZebraPuzzle (Resident (..), Solution (..), solve) where

import Control.Monad (guard)
import Data.List (permutations, zip5)

data Solution = Solution
  { waterDrinker :: Resident
  , zebraOwner :: Resident
  }
  deriving (Eq, Show)

data Resident = Englishman | Spaniard | Ukrainian | Norwegian | Japanese deriving (Eq, Show, Enum)
data Color = Blue | Red | Ivory | Green | Yellow deriving (Eq, Enum)
data Drink = Tea | Coffee | Milk | Orange | Water deriving (Eq, Enum)
data Animal = Dog | Snails | Horse | Fox | Zebra deriving (Eq, Enum)
data Cigarette = Parliaments | OldGold | Kools | Chesterfields | LuckyStrike deriving (Eq, Enum)

solve :: Solution
solve = Solution{waterDrinker = whoDrinksWater, zebraOwner = whoOwnsZebra}
 where
  whoDrinksWater = head [resident | (_, resident, _, Water, _) <- solution]
  whoOwnsZebra = head [resident | (_, resident, Zebra, _, _) <- solution]
  solution = head $ do
    nationalities <- permutations [Englishman ..]
    guard $ head nationalities == Norwegian
    colors <- permutations [Blue ..]
    guard $ (Ivory, Green) `elem` zip colors (tail colors)
    guard $ (Englishman, Red) `elem` zip nationalities colors
    guard $ (Norwegian, Blue) `elem` zip nationalities (tail colors) || (Blue, Norwegian) `elem` zip colors (tail nationalities)
    drinks <- permutations [Tea ..]
    guard $ (Milk, 2) `elem` zip drinks ([0, 1, 2, 3, 4] :: [Int])
    guard $ (Coffee, Green) `elem` zip drinks colors
    guard $ (Ukrainian, Tea) `elem` zip nationalities drinks
    animals <- permutations [Dog ..]
    guard $ (Spaniard, Dog) `elem` zip nationalities animals
    cigarettes <- permutations [Parliaments ..]
    guard $ (OldGold, Snails) `elem` zip cigarettes animals
    guard $ (Kools, Yellow) `elem` zip cigarettes colors
    guard $ (LuckyStrike, Orange) `elem` zip cigarettes drinks
    guard $ (Japanese, Parliaments) `elem` zip nationalities cigarettes
    return $ zip5 colors nationalities animals drinks cigarettes

File: solutions/prolog/zebra-puzzle/1/zebra_puzzle.pl
zebra_owner(Owner) :-
	houses(Hs),
	member(h(Owner,zebra,_,_,_), Hs).

water_drinker(Drinker) :-
	houses(Hs),
	member(h(Drinker,_,_,water,_), Hs).

houses(Houses) :-
	length(Houses, 5),                                            %  1
	member(h(english,_,_,_,red), Houses),                         %  2
	member(h(spanish,dog,_,_,_), Houses),                         %  3
	member(h(_,_,_,coffee,green), Houses),                        %  4
	member(h(ukrainian,_,_,tea,_), Houses),                       %  5
	adjacent(h(_,_,_,_,green), h(_,_,_,_,white), Houses),             %  6
	member(h(_,snake,winston,_,_), Houses),                       %  7
	member(h(_,_,kool,_,yellow), Houses),                         %  8
	Houses = [_,_,h(_,_,_,milk,_),_,_],                           %  9
	Houses = [h(norwegian,_,_,_,_)|_],                            % 10
	adjacent(h(_,fox,_,_,_), h(_,_,chesterfield,_,_), Houses),        % 11
	adjacent(h(_,_,kool,_,_), h(_,horse,_,_,_), Houses),              % 12
	member(h(_,_,lucky,juice,_), Houses),                         % 13
	member(h(japanese,_,kent,_,_), Houses),                       % 14
	adjacent(h(norwegian,_,_,_,_), h(_,_,_,_,blue), Houses),          % 15
	member(h(_,_,_,water,_), Houses),		% one of them drinks water
	member(h(_,zebra,_,_,_), Houses).		% one of them owns a zebra

adjacent(A, B, Ls) :- append(_, [A,B|_], Ls).
adjacent(A, B, Ls) :- append(_, [B,A|_], Ls).
